src/algebra/algebra/basic.lean

```diff
@@ -115,6 +115,36 @@ class algebra (R : Type u) (A : Type v) [comm_semiring R] [semiring A]
   extends has_scalar R A, R →+* A :=
 (commutes' : ∀ r x, to_fun r * x = x * to_fun r)
 (smul_def' : ∀ r x, r • x = to_fun r * x)
+
+/--
+A (non-unital, non-associative) `R`-algebra is a non-unital, non-associative, semiring `A` equipped
+with a map into its centroid `R → Z(A)`.
+-/
+class non_unital_non_assoc_algebra (R : Type u) (A : Type v) [comm_semiring R]
+  [non_unital_non_assoc_semiring A] extends has_scalar R A, R →+* (add_monoid.End A) :=
```

Comment:
To be this just feels like a convoluted way to write
```suggestion
  [non_unital_non_assoc_semiring A] extends module R A :=
```

Reply:
Personally I think of an algebra as being a bilinear map on a vector space (or more generally a module). From the comments you wrote at the top of this file https://github.com/leanprover-community/mathlib/blob/master/src/algebra/algebra/basic.lean#L48-L95 - particularly https://github.com/leanprover-community/mathlib/blame/master/src/algebra/algebra/basic.lean#L94-L95 - I understood that not to be the preferred way of thinking about algebras in mathlib?

Reply:
You can read the typeclasses as:

* `[non_unital_non_assoc_semiring A]`: a binary map  `(*)` preserving addition
* `[module R A]`: on a vector space
* `[is_scalar_tower R A A] [smul_comm_class R A A]`: that is bilinear